File: usefulmethodx/src/main/java/com/acdev/usefulmethodx/ShimmerRecycler.java
package com.acdev.usefulmethodx;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.AttributeSet;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.mikelau.views.shimmer.BindViewHolderPlugin;
import com.mikelau.views.shimmer.ShimmerAdapter;

public class ShimmerRecycler extends RecyclerView {

    public enum LayoutMangerType {
        LINEAR_VERTICAL, LINEAR_HORIZONTAL, GRID
    }

    private Adapter mActualAdapter;
    private ShimmerAdapter mShimmerAdapter;

    private LayoutManager mShimmerLayoutManager;
    private LayoutManager mActualLayoutManager;
    private LayoutMangerType mLayoutMangerType;

    private boolean mCanScroll;
    private int mLayoutReference;
    private int mGridCount;

    private int mShimmerAngle;
    private int mShimmerColor;
    private int mShimmerDuration;
    private float mShimmerMaskWidth;
    private boolean isAnimationReversed;
    private Drawable mShimmerItemBackground;

    public ShimmerRecycler(Context context) {
        super(context);
        init(context, null);
    }

    public ShimmerRecycler(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public ShimmerRecycler(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        mShimmerAdapter = new ShimmerAdapter();
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.ShimmerRecyclerView, 0, 0);

        try {
            setDemoLayoutReference(a.getResourceId(R.styleable.ShimmerRecyclerView_shimmer_demo_layout, R.layout.layout_sample_view));
            setDemoChildCount(a.getInteger(R.styleable.ShimmerRecyclerView_shimmer_demo_child_count, 10));
            setGridChildCount(a.getInteger(R.styleable.ShimmerRecyclerView_shimmer_demo_grid_child_count, 2));

            final int value = a.getInteger(R.styleable.ShimmerRecyclerView_shimmer_demo_layout_manager_type, 0);
            switch (value) {
                case 0:
                    setDemoLayoutManager(LayoutMangerType.LINEAR_VERTICAL);
                    break;
                case 1:
                    setDemoLayoutManager(LayoutMangerType.LINEAR_HORIZONTAL);
                    break;
                case 2:
                    setDemoLayoutManager(LayoutMangerType.GRID);
                    break;
                default:
                    throw new IllegalArgumentException("This value for layout manager is not valid!");
            }

            mShimmerAngle = a.getInteger(R.styleable.ShimmerRecyclerView_shimmer_demo_angle, 0);
            mShimmerColor = a.getColor(R.styleable.ShimmerRecyclerView_shimmer_demo_shimmer_color, getColor(R.color.default_shimmer_color));
            mShimmerItemBackground = a.getDrawable(R.styleable.ShimmerRecyclerView_shimmer_demo_view_holder_item_background);
            mShimmerDuration = a.getInteger(R.styleable.ShimmerRecyclerView_shimmer_demo_duration, 1500);
            mShimmerMaskWidth = a.getFloat(R.styleable.ShimmerRecyclerView_shimmer_demo_mask_width, 0.5f);
            isAnimationReversed = a.getBoolean(R.styleable.ShimmerRecyclerView_shimmer_demo_reverse_animation, false);
        } finally {
            a.recycle();
        }

        setupShimmerProperties();
    }

    private void setupShimmerProperties() {
        mShimmerAdapter.setShimmerAngle(mShimmerAngle);
        mShimmerAdapter.setShimmerColor(mShimmerColor);
        mShimmerAdapter.setShimmerMaskWidth(mShimmerMaskWidth);
        mShimmerAdapter.setShimmerItemBackground(mShimmerItemBackground);
        mShimmerAdapter.setShimmerDuration(mShimmerDuration);
        mShimmerAdapter.setAnimationReversed(isAnimationReversed);
        showShimmerAdapter();
    }

    public void setBindViewHolderPlugin(BindViewHolderPlugin plugin) {
        mShimmerAdapter.setBindViewHolderPlugin(plugin);
    }

    public void setGridChildCount(int count) {
        mGridCount = count;
    }

    public void setDemoLayoutManager(LayoutMangerType type) {
        mLayoutMangerType = type;
    }

    public void setDemoChildCount(int count) {
        mShimmerAdapter.setMinItemCount(count);
    }

    public void setDemoShimmerDuration(int duration) {
        mShimmerAdapter.setShimmerDuration(duration);
    }

    public void setDemoShimmerMaskWidth(float maskWidth) {
        mShimmerAdapter.setShimmerMaskWidth(maskWidth);
    }

    public void showShimmerAdapter() {
        mCanScroll = false;

        if (mShimmerLayoutManager == null) {
            initShimmerManager();
        }

        setLayoutManager(mShimmerLayoutManager);
        setAdapter(mShimmerAdapter);
    }

    public void hideShimmerAdapter() {
        mCanScroll = true;
        setLayoutManager(mActualLayoutManager);
        setAdapter(mActualAdapter);
    }

    public void setLayoutManager(LayoutManager manager) {
        if (manager == null) {
            mActualLayoutManager = null;
        } else if (manager != mShimmerLayoutManager) {
            mActualLayoutManager = manager;
        }

        super.setLayoutManager(manager);
    }

    public void setAdapter(Adapter adapter) {
        if (adapter == null) {
            mActualAdapter = null;
        } else if (adapter != mShimmerAdapter) {
            mActualAdapter = adapter;
        }

        super.setAdapter(adapter);
    }

    public Adapter getActualAdapter() {
        return mActualAdapter;
    }

    public Adapter getShimmerAdapter() {
        return mShimmerAdapter;
    }

    public int getLayoutReference() {
        return mLayoutReference;
    }

    public void setDemoLayoutReference(int mLayoutReference) {
        this.mLayoutReference = mLayoutReference;
        mShimmerAdapter.setLayoutReference(getLayoutReference());
    }

    private void initShimmerManager() {
        switch (mLayoutMangerType) {
            case LINEAR_VERTICAL:
                mShimmerLayoutManager = new LinearLayoutManager(getContext()) {
                    public boolean canScrollVertically() {
                        return mCanScroll;
                    }
                };
                break;
            case LINEAR_HORIZONTAL:
                mShimmerLayoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false) {
                    public boolean canScrollHorizontally() {
                        return mCanScroll;
                    }
                };
                break;
            case GRID:
                mShimmerLayoutManager = new GridLayoutManager(getContext(), mGridCount) {
                    public boolean canScrollVertically() {
                        return mCanScroll;
                    }
                };
                break;
        }
    }

    private int getColor(int id) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            return getContext().getColor(id);
        } else {
            return getResources().getColor(id);
        }
    }
}

File: commonFunction/src/main/java/com/acdev/commonFunction/ShimmerRecycler.kt
package com.acdev.commonFunction

import android.content.Context
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.mikelau.views.shimmer.BindViewHolderPlugin
import com.mikelau.views.shimmer.ShimmerAdapter
import kotlinx.android.synthetic.main.layout_empty.view.*

class ShimmerRecycler : RecyclerView {

    enum class LayoutMangerType { LINEAR_VERTICAL, LINEAR_HORIZONTAL, GRID }

    private var mActualAdapter: Adapter<*>? = null
    private var mShimmerAdapter: ShimmerAdapter? = null
    private var mShimmerLayoutManager: LayoutManager? = null
    private var mActualLayoutManager: LayoutManager? = null
    private var mLayoutMangerType: LayoutMangerType? = null
    private var mCanScroll = false
    private var mLayoutReference = 0
    private var mGridCount = 0
    private var mShimmerAngle = 0
    private var mShimmerColor = 0
    private var mShimmerDuration = 0
    private var mShimmerMaskWidth = 0f
    private var isAnimationReversed = false
    private var mShimmerItemBackground: Drawable? = null
    private var emptyView: View? = null

    constructor(context: Context) : super(context) { init(context, null) }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) { init(context, attrs) }

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) { init(context, attrs) }

    private fun init(context: Context, attrs: AttributeSet?) {
        mShimmerAdapter = ShimmerAdapter()
        val a = context.obtainStyledAttributes(attrs, R.styleable.ShimmerRecycler, 0, 0)
        try {
            layoutReference(a.getResourceId(R.styleable.ShimmerRecycler_layout, R.layout.layout_sample_view))
            childCount = a.getInteger(R.styleable.ShimmerRecycler_child_count, 10)
            a.getResourceId(R.styleable.ShimmerRecycler_empty_layout, 0)
            setGridChildCount(a.getInteger(R.styleable.ShimmerRecycler_grid_child_count, 2))
            when (a.getInteger(R.styleable.ShimmerRecycler_layout_manager, 0)) {
                0 -> layoutManager(LayoutMangerType.LINEAR_VERTICAL)
                1 -> layoutManager(LayoutMangerType.LINEAR_HORIZONTAL)
                2 -> layoutManager(LayoutMangerType.GRID)
                else -> throw IllegalArgumentException("This value for layout manager is not valid!")
            }
            mShimmerAngle = a.getInteger(R.styleable.ShimmerRecycler_angle, 0)
            mShimmerColor = a.getColor(R.styleable.ShimmerRecycler_color, getColor(R.color.color))
            mShimmerItemBackground = a.getDrawable(R.styleable.ShimmerRecycler_item_background)
            mShimmerDuration = a.getInteger(R.styleable.ShimmerRecycler_duration, 1500)
            mShimmerMaskWidth = a.getFloat(R.styleable.ShimmerRecycler_mask_width, 0.5f)
            isAnimationReversed = a.getBoolean(R.styleable.ShimmerRecycler_reverse_animation, false)
        } finally { a.recycle() }
        setupShimmerProperties()
    }

    private fun setupShimmerProperties() {
        mShimmerAdapter!!.setShimmerAngle(mShimmerAngle)
        mShimmerAdapter!!.setShimmerColor(mShimmerColor)
        mShimmerAdapter!!.setShimmerMaskWidth(mShimmerMaskWidth)
        mShimmerAdapter!!.setShimmerItemBackground(mShimmerItemBackground)
        mShimmerAdapter!!.setShimmerDuration(mShimmerDuration)
        mShimmerAdapter!!.setAnimationReversed(isAnimationReversed)
        showShimmer()
    }

    fun setBindViewHolderPlugin(plugin: BindViewHolderPlugin?) {
        mShimmerAdapter!!.setBindViewHolderPlugin(plugin)
    }

    fun setGridChildCount(count: Int) {
        mGridCount = count
    }

    fun layoutManager(type: LayoutMangerType?) {
        mLayoutMangerType = type
    }

    fun setChildCount(count: Int) {
        mShimmerAdapter!!.setMinItemCount(count)
    }

    fun duration(duration: Int) {
        mShimmerAdapter!!.setShimmerDuration(duration)
    }

     fun maskWidth(maskWidth: Float) {
        mShimmerAdapter!!.setShimmerMaskWidth(maskWidth)
    }

    fun showShimmer() {
        mCanScroll = false
        if (mShimmerLayoutManager == null) initShimmerManager()
        layoutManager = mShimmerLayoutManager
        adapter = mShimmerAdapter
    }

    fun hideShimmer() {
        mCanScroll = true
        layoutManager = mActualLayoutManager
        adapter = mActualAdapter
    }

    fun Context.setLayout(recyclerAdapter: Adapter<*>?, spanCount: Int) {
        val layoutManagerx: LayoutManager = GridLayoutManager(this, spanCount)
        layoutManager = layoutManagerx
        adapter = recyclerAdapter
        adapter?.notifyDataSetChanged()
    }

    override fun onMeasure(widthSpec: Int, heightSpec: Int) {
        val view = emptyView
        if (view != null && adapter?.itemCount == 0) {
            view.measure(widthSpec, heightSpec)
            val width = view.measuredWidth
            val height = view.measuredHeight
            setMeasuredDimension(width, height)
            return
        }

        super.onMeasure(widthSpec, heightSpec)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
        emptyView?.layout(l, t, r, b)
    }

    override fun onDraw(c: Canvas) {
        super.onDraw(c)
        if (adapter?.itemCount == 0) {
            emptyView?.draw(c)
        }
    }


    @SuppressWarnings("unused")
    fun setEmptyView(view: View?) {
        emptyView = view

        invalidate()
    }

    /**
     * Set a view to show if the adapter is empty
     *
     * @param resId empty view's layout id
     */
    @SuppressWarnings("unused")
    fun setEmptyView(@LayoutRes resId: Int) {
        if (resId == 0) return

        val view = LayoutInflater.from(context).inflate(resId, null)
        setEmptyView(view)
    }


    fun setEmptyImage(layout: Int) {
        icon.setImageResource(layout)
    }

    fun setEmptyText(string: String) {
        text.text = string
    }

    fun setEmptyText(string: Int) {
        text.text = this.resources.getString(string)
    }

    override fun setLayoutManager(manager: LayoutManager?) {
        if (manager == null) mActualLayoutManager = null
        else if (manager !== mShimmerLayoutManager) mActualLayoutManager = manager
        super.setLayoutManager(manager)
    }

    override fun setAdapter(adapter: Adapter<*>?) {
        if (adapter == null) mActualAdapter = null
        else if (adapter !== mShimmerAdapter) mActualAdapter = adapter
        super.setAdapter(adapter)
    }

    fun getActualAdapter(): Adapter<*>? {
        return mActualAdapter
    }

    fun getShimmerAdapter(): Adapter<*>? {
        return mShimmerAdapter
    }

    fun getLayoutReference(): Int {
        return mLayoutReference
    }

    fun layoutReference(mLayoutReference: Int) {
        this.mLayoutReference = mLayoutReference
        mShimmerAdapter!!.setLayoutReference(getLayoutReference())
    }

    private fun initShimmerManager() {
        when (mLayoutMangerType) {
            LayoutMangerType.LINEAR_VERTICAL -> mShimmerLayoutManager = object : LinearLayoutManager(context) {
                    override fun canScrollVertically(): Boolean { return mCanScroll } }
            LayoutMangerType.LINEAR_HORIZONTAL -> mShimmerLayoutManager = object : LinearLayoutManager(
                    context, HORIZONTAL, false) {
                    override fun canScrollHorizontally(): Boolean { return mCanScroll } }
            LayoutMangerType.GRID -> mShimmerLayoutManager = object : GridLayoutManager(context, mGridCount) {
                    override fun canScrollVertically(): Boolean { return mCanScroll } }
        }
    }

    private fun getColor(id: Int): Int {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) context.getColor(id)
        else resources.getColor(id)
    }
}